helix-term/src/ui/picker.rs

```diff
@@ -949,6 +877,108 @@ impl<T: Item> Drop for Picker<T> {
 
 type PickerCallback<T> = Box<dyn Fn(&mut Context, &T, Action)>;
 
+struct PreviewHighlightHandler<T: ui::menu::Item> {
+    phantom_data: std::marker::PhantomData<T>,
+}
+
+impl<T: ui::menu::Item> Default for PreviewHighlightHandler<T> {
+    fn default() -> Self {
+        Self {
+            phantom_data: Default::default(),
+        }
+    }
+}
+
+impl<T: ui::menu::Item> AsyncHook for PreviewHighlightHandler<T> {
+    type Event = ();
+
+    fn handle_event(
+        &mut self,
+        _event: Self::Event,
+        _timeout: Option<tokio::time::Instant>,
+    ) -> Option<tokio::time::Instant> {
+        Some(Instant::now() + Duration::from_millis(250))
+    }
+
+    fn finish_debounce(&mut self) {
+        crate::job::dispatch_blocking(|editor, compositor| {
+            let picker = match compositor.find::<Overlay<Picker<T>>>() {
+                Some(Overlay { content, .. }) => content,
+                None => match compositor.find::<Overlay<DynamicPicker<T>>>() {
+                    Some(Overlay { content, .. }) => &mut content.file_picker,
+                    None => return,
+                },
+            };
+
+            let Some((current_file, _)) = picker.current_file(editor) else { return };
+            // Try to find a document in the cache
+            let doc = match &current_file {
+                PathOrId::Id(doc_id) => doc_mut!(editor, doc_id),
+                PathOrId::Path(path) => match picker.preview_cache.get_mut(path) {
+                    Some(CachedPreview::Document(ref mut doc)) => doc,
+                    _ => return,
+                },
+            };
+
+            if doc.language_config().is_some() {
+                return;
+            }
+
+            let Some(language_config) = doc.detect_language_config(&editor.syn_loader.load()) else { return };
+            doc.language = Some(language_config.clone());
+            let text = doc.text().clone();
+            let loader = editor.syn_loader.clone();
+
+            tokio::spawn(async move {
```

Comment:
this still needs to be spawn_blocking. We shouldn't be CPU intense stuff (which parsing definitely is) in an async job. Using job::dispatch_blocking isn't much of an issue. In theory you could also use block in place but that probably has more overhead than just using dispatch_blocking (dispatch almost never blocks anyway)